File: ip2cc.rb
VALID_IP = Regexp.new('^'+(['(?:(?:2[0-4]|1\d|[1-9])?\d|25[0-5])']*4).join('\.')+'$')

class IP2CC
  
  def initialize(filename)
    @filename = filename
  end
  
  def [](ip, opts={})
    raise 'invalid ip' unless ip =~ VALID_IP
    open(@filename) do |f|
      offset = 0
      ip.split('.').each do |part|
       start = offset + part.to_i * 4
       f.seek start
       value = f.read 4
       if value.slice(0, 2) == "\xFF\xFF" then
         if value.slice(2, 2) == "\x00\x00" then
           raise 'not found' if opts[:raise] 
           return nil
         end
         return value.slice(2, 2)
       end
       offset = value.unpack('N')[0]
      end
    end
    raise 'broken db'
  end
  
end

if __FILE__ == $0 then
  db = IP2CC.new 'ip2cc.db'
  if ARGV.length == 1 then
    puts db[ARGV[0]]
  else
    ARGV.each{ |ip| puts "#{ip} #{db[ip]}" }
  end
end

File: ip2cc.php
<?php

class IP2CC implements ArrayAccess {
  
  public function __construct($filename) {
    $this->filename = $filename;
  }
  
  public function offsetGet($ip) {
    $fp = fopen($this->filename, 'rb');
    $parts = explode('.', $ip);
    $offset = 0;
    foreach($parts as $part) {
      $start = $offset + intval($part) * 4;
      fseek($fp, $start);
      $value = fread($fp, 4);
      if (substr($value, 0, 2) == "\xFF\xFF") {
        $res = substr($value, 2, 2);
        if ($res == "\x00\x00") {
          return;
        }
        return $res;
      }
      $chunks = unpack('N', $value);
      $offset = $chunks[1];
    }
    throw new Exception('broken db');
  }
  
  public function offsetExists($foo) {
  }
  
  public function offsetUnset($bar) {
  }
  
  public function offsetSet($spam, $eggs) {
  }
  
}

if (count($argv) > 1 && realpath($argv[0]) == __file__ ){
  
  $a = new IP2CC('ip2cc.db');
  $ip = $argv[1];
  print "$a[$ip]\n";
  
}

